exporters/committime/collector_base.py

```diff
@@ -120,14 +121,53 @@ def generate_metrics(
             watched_namespaces = [
                 namespace.metadata.name for namespace in v1_namespaces.get().items
             ]
+        if watched_namespaces:
+            # Case where comma separated list was provided as set {}
+            watched_namespaces = list(watched_namespaces)
```

Comment:
ensure the watched_namespaces are type of "list" and not type of "set". While debugging saw it's a set, see:

```
$ export NAMESPACES=first,second
(breakpoint on line 125)
(Pdb) print(watched_namespaces)
{'first', 'second'}
(Pdb) type(watched_namespaces)
<class 'set'>

on lin 127:
(Pdb) print(watched_namespaces)
['first', 'second']
(Pdb) type(watched_namespaces)
<class 'list'>
```

Reply:
All it needs to be is an iterable of strings. I made it a set because otherwise, using `NAMESPACES=first,second,first` will make it query everything in namespace `first` twice.

Reply:
Sure, but if it's set and we convert it here to list then it's ok as set already filters out the duplicates *and* we return expected type. What do you think?

Reply:
Ok in the newest revision added overload. The list is when no NAMESPACE is defined in env var, when something is specified then it's set.

Reply:
Right, but it also doesn't _need_ to be a list. We can change the return type annotation. Simply put:

```patch
diff --git a/exporters/committime/collector_base.py b/exporters/committime/collector_base.py
index a6fab57..ef9de6e 100644
--- a/exporters/committime/collector_base.py
+++ b/exporters/committime/collector_base.py
@@ -20,7 +20,7 @@ from __future__ import annotations
 import logging
 import re
 from abc import abstractmethod
-from typing import Iterable, Optional, overload
+from typing import Iterable, Optional
 
 from jsonpath_ng import parse
 from prometheus_client.core import GaugeMetricFamily
@@ -57,7 +57,7 @@ class AbstractCommitCollector(pelorus.AbstractPelorusExporter):
         kube_client,
         username: str,
         token: str,
-        namespaces: list[str],
+        namespaces: set[str],
         apps,
         collector_name,
         timedate_format,
@@ -84,10 +84,7 @@ class AbstractCommitCollector(pelorus.AbstractPelorusExporter):
             labels=["namespace", "app", "commit", "image_sha"],
         )
 
-        if self._namespaces is not None:
-            commit_metrics = self.generate_metrics(self._namespaces)
-        else:
-            commit_metrics = self.generate_metrics()
+        commit_metrics = self.generate_metrics()
 
         for my_metric in commit_metrics:
             logging.info(
@@ -112,28 +109,22 @@ class AbstractCommitCollector(pelorus.AbstractPelorusExporter):
             )
             yield commit_metric
 
-    @overload
-    def _get_watched_namespaces(self, watched_namespaces: list[str] = None) -> set[str]:
-        ...
-
-    def _get_watched_namespaces(
-        self, watched_namespaces: list[str] = None
-    ) -> list[str]:
+    def _get_watched_namespaces(self) -> set[str]:
+        watched_namespaces = self._namespaces
         if not watched_namespaces:
             logging.info("No namespaces specified, watching all namespaces")
             v1_namespaces = self._kube_client.resources.get(
                 api_version="v1", kind="Namespace"
             )
-            watched_namespaces = [
+            watched_namespaces = {
                 namespace.metadata.name for namespace in v1_namespaces.get().items
-            ]
+            }
         logging.info("Watching namespaces: %s" % (watched_namespaces))
         return watched_namespaces
 
     def _get_openshift_obj_by_app(
         self, openshift_obj: str, app_label: str
-    ) -> Optional(dict):
-        apps = []
+    ) -> Optional[dict]:
 
         # use a jsonpath expression to find all values for the app label
         jsonpath_str = "$['items'][*]['metadata']['labels']['" + str(app_label) + "']"
@@ -142,13 +133,11 @@ class AbstractCommitCollector(pelorus.AbstractPelorusExporter):
 
         found = jsonpath_expr.find(openshift_obj)
 
-        apps = [match.value for match in found]
+        apps = {match.value for match in found}
 
         if not apps:
             return None
 
-        # remove duplicates
-        apps = list(dict.fromkeys(apps))
         items_by_app = {}
 
         for app in apps:
@@ -160,13 +149,11 @@ class AbstractCommitCollector(pelorus.AbstractPelorusExporter):
 
         return items_by_app
 
-    def generate_metrics(
-        self, watched_namespaces: Optional[list[str]] = None
-    ) -> Iterable[CommitMetric]:
+    def generate_metrics(self) -> Iterable[CommitMetric]:
         """Method called by the collect to create a list of metrics to publish"""
         # This will loop and look at OCP builds (calls get_git_commit_time)
 
-        watched_namespaces = self._get_watched_namespaces(watched_namespaces)
+        watched_namespaces = self._get_watched_namespaces()
 
         # Initialize metrics list
         metrics = []
```

We're taking lists, removing duplicates, and then turning them back into lists, only to iterate over them later. We can just use a set directly.

Reply:
Ok I am cool with that, applied patch, thanks.